File: tools/hod2lib/degraded.py
"""What the exporter could not read, and carried on without.

Thirty-odd places in this package answer a failure with an empty result:
``except Exception: return {}``, ``return []``, ``cache[stem] = ([], None)``.
Every one of them is deliberate -- an install missing one `pol/` file should
still produce a bundle, and refusing to export stage 2 because a single
damaged prop model will not parse would be worse than exporting without it.

What was wrong is that they were **silent**. A parser regression anywhere
under here produced a valid bundle with zero characters, exit code 0, and no
message; the player then rendered an empty stage, which looks exactly like a
gameplay bug. `docs/PLAN.md` P6 named "a silent exporter regression" as the
thing nothing in this repository could catch, and this is why.

So the shape stays and the silence goes. A site that swallows says so:

    try:
        prog = scriptlib.load(stage)
    except Exception as exc:
        degraded.note("the stage's event script", "no props", exc)
        return [], []

which prints one line to stderr and appends a record. `build_stage` drains the
records into the stage's manifest entry, so a bundle carries the list of what
is missing from it, and **`export_player.py` exits non-zero if that list is
not empty.** There is no flag for that and there should not be: a `--strict`
nobody passes is a check that never fires, which is the finding this module
answers, one level up. The files are still written, because an incomplete
bundle is usually what you want to look at while finding out why.

**`where` is taken from the caller's frame**, not passed in. A hand-written
location string is a second name for the same thing and it goes stale the
first time a function moves.
"""
from __future__ import annotations

import sys
from dataclasses import asdict, dataclass


@dataclass(frozen=True)
class Degradation:
    """One thing the export could not read, and what was given up for it."""

    #: ``module.function`` of the call site, from the caller's frame.
    where: str
    #: What was being read, in the words of whoever wrote the site.
    what: str
    #: What the bundle is missing as a result. The half that matters.
    lost: str
    #: ``TypeName: message``. Not a traceback -- the site is the location.
    error: str

    def line(self) -> str:
        return f"{self.where}: could not read {self.what} ({self.error}) — {self.lost}"


_log: list[Degradation] = []
#: ``(where, exception type)`` already printed. A cache miss inside a loop over
#: two hundred asset slots is one story, not two hundred lines of it; every
#: occurrence is still recorded, only the printing is folded.
_printed: set[tuple[str, str]] = set()


def note(what: str, lost: str, exc: BaseException) -> None:
    """Record that *what* could not be read, and that *lost* is the price."""
    frame = sys._getframe(1)                                    # noqa: SLF001
    where = f"{frame.f_globals.get('__name__', '?')}.{frame.f_code.co_name}"
    rec = Degradation(where=where, what=what, lost=lost,
                      error=f"{type(exc).__name__}: {exc}")
    _log.append(rec)
    key = (where, type(exc).__name__)
    if key not in _printed:
        _printed.add(key)
        print(f"  warning: {rec.line()}", file=sys.stderr)


def reset() -> None:
    """Start a fresh count. `build_stage` calls this so a count is per stage."""
    _log.clear()
    _printed.clear()


def drain() -> list[dict]:
    """Everything recorded since `reset`, as plain dicts, and start again."""
    out = [asdict(r) for r in _log]
    reset()
    return out

File: web/src/hod2lib/degraded.ts
/**
 * What the exporter could not read, and carried on without.
 *
 * A port of `tools/hod2lib/degraded.py`, and the reasoning there is the whole
 * story: thirty-odd places in this package answer a failure with an empty
 * result, every one of them deliberate, and what was wrong is that they were
 * silent. A site that swallows says so, `buildStage` drains the records into
 * the stage's manifest entry, and the export exits non-zero if the list is not
 * empty.
 *
 * **One divergence, and it is in the `where` field.** Python takes it from the
 * caller's frame, `sys._getframe(1)`, so a moved function cannot leave a stale
 * location behind. JavaScript has no equivalent that survives bundling:
 * esbuild renames functions and the browser build has no file names at all. So
 * the call site passes it, spelled exactly as Python would have derived it
 * (`hod2lib.nl1.parse`), and `tools/compare_bundles.py` compares the strings.
 * A hand-written location can go stale, which is the cost; a location that
 * reads `t.a` after minification is worse.
 */

export interface Degradation {
  /** `module.function` of the call site. */
  where: string;
  /** What was being read, in the words of whoever wrote the site. */
  what: string;
  /** What the bundle is missing as a result. The half that matters. */
  lost: string;
  /** `TypeName: message`. Not a stack; the site is the location. */
  error: string;
}

/** The em dash is the reference implementation's; the strings are compared. */
export function line(r: Degradation): string {
  return `${r.where}: could not read ${r.what} (${r.error}) — ${r.lost}`;
}

const log: Degradation[] = [];
/**
 * `(where, error type)` already printed. A cache miss inside a loop over two
 * hundred asset slots is one story, not two hundred lines of it; every
 * occurrence is still recorded, only the printing is folded.
 */
const printed = new Set<string>();

let warn: (text: string) => void = () => {};

/** Where a warning goes. The CLI sends it to stderr; the page, to the feed. */
export function setWarningSink(fn: (text: string) => void): void {
  warn = fn;
}

/**
 * `TypeName: message`, the way Python renders an exception in a record.
 *
 * An `Error` subclass keeps its constructor name through esbuild because the
 * classes in this package are declared rather than generated; anything thrown
 * that is not an `Error` is reported as its own text, which is what Python's
 * `str(exc)` does for a value with no message of its own.
 */
export function describe(exc: unknown): string {
  if (exc instanceof Error) return `${exc.name}: ${exc.message}`;
  return String(exc);
}

/** Record that *what* could not be read, and that *lost* is the price. */
export function note(where: string, what: string, lost: string,
                     exc: unknown): void {
  const rec: Degradation = { where, what, lost, error: describe(exc) };
  log.push(rec);
  const key = `${where} ${rec.error.split(":")[0]}`;
  if (!printed.has(key)) {
    printed.add(key);
    warn(`  warning: ${line(rec)}`);
  }
}

/** Start a fresh count. `buildStage` calls this so a count is per stage. */
export function reset(): void {
  log.length = 0;
  printed.clear();
}

/** Everything recorded since {@link reset}, and start again. */
export function drain(): Degradation[] {
  const out = log.slice();
  reset();
  return out;
}
